service/report-task.cpp

```diff
@@ -89,7 +101,159 @@ int run_event_chain(struct run_event_state *run_state, const char *dump_dir_name
+                                 GDBusMethodInvocation *invocation,
+                                 gpointer               user_data)
+{
+    g_object_set_data(G_OBJECT(object), "handled", GINT_TO_POINTER(1));
+
+    return TRUE;
+}
+
+static ReportDbusTaskPrompt *
+report_task_emit_prompt(ReportTask *self,
+                        const char *message,
+                        PromptType  type)
+{
+    g_autoptr(GDBusObjectSkeleton) prompt_skeleton = NULL;
+    ReportDbusTaskPrompt *prompt_interface = NULL;
+    const char *object_path;
+
+    prompt_skeleton = g_dbus_object_skeleton_new(REPORTD_DBUS_TASK_PROMPT_PATH);
+    prompt_interface = report_dbus_task_prompt_skeleton_new();
+
+    g_dbus_object_skeleton_add_interface(prompt_skeleton,
+                                         G_DBUS_INTERFACE_SKELETON(prompt_interface));
+
+    ReportDaemon::inst().register_object(prompt_skeleton);
+
+    object_path = g_dbus_object_get_object_path(G_DBUS_OBJECT(prompt_skeleton));
+
+    g_signal_connect(prompt_interface, "handle-commit",
+                     G_CALLBACK(report_task_prompt_handle_commit), NULL);
+
+    report_dbus_task_emit_prompt(self->task_iface, object_path, message, type);
+
+    while (g_object_get_data(G_OBJECT(prompt_interface), "handled") != GINT_TO_POINTER(1))
+    {
+        g_main_context_iteration(NULL, TRUE);
+    }
+
+    g_object_set_data(G_OBJECT(prompt_interface), "handled", NULL);
+
+    ReportDaemon::inst().unregister_object(G_DBUS_OBJECT(prompt_skeleton));
+
+    return prompt_interface;
+}
+
+static char *
+report_task_ask_callback (const char *msg,
+                          void       *interaction_param)
+{
+    ReportTask *self;
+    g_autoptr(ReportDbusTaskPrompt) prompt_interface = NULL;
+    const char *input;
+
+    self = REPORT_TASK(interaction_param);
+    prompt_interface = report_task_emit_prompt(self, msg, ASK);
+    input = report_dbus_task_prompt_get_input(prompt_interface);
+
+    return g_strdup(input);
+}
+
+static int
+report_task_ask_yes_no_callback(const char *msg,
+                                void       *interaction_param)
+{
+    ReportTask *self;
+    g_autoptr(ReportDbusTaskPrompt) prompt_interface = NULL;
+
+    self = REPORT_TASK(interaction_param);
+    prompt_interface = report_task_emit_prompt(self, msg, ASK_YES_NO);
+
+    return report_dbus_task_prompt_get_response(prompt_interface);
+}
+
+static int
+report_task_ask_yes_no_yesforever_callback(const char *key,
+                                           const char *msg,
+                                           void       *interaction_param)
+{
+    const char *value;
+    ReportTask *self;
+    g_autoptr(ReportDbusTaskPrompt) prompt_interface = NULL;
+    bool response;
+    bool remember;
+
+    value = get_user_setting(key);
+    /* The following is replicating the madness inside libreport, where
+     * “no” means “yes, forever”, and “yes” means nothing, really.
```

Comment:
https://github.com/abrt/libreport/commit/2ed85dae0d8832a36102116be0480eba771846f6
```
    - somebody can be confused by the logic but it was originally
      implemented in report-gtk ("Don't ask me again" check box)
```

I wanted to keep the new implementation backward compatible.

Perhaps, I should ignore backward compatibility and let users rage about ABRT asking them even thought they checked the "Do not ask" checkbox. But I was afraid because ABRT's reputation wasn't good and I didn't want put more fuel into the fire (I am not as brave developer as GNOME developers).

Reply:
I guess this is a testament to how much of a problem negative phrasing like that is (https://developer.gnome.org/hig/stable/check-boxes.html.en). Seems not much more than a slight annoyance that one would have to answer the prompt again.